src/main/java/ch/njol/skript/effects/EffCancelEvent.java

```diff
@@ -43,65 +32,87 @@
+import org.bukkit.event.player.PlayerDropItemEvent;
+import org.bukkit.event.player.PlayerInteractEvent;
+import org.bukkit.event.player.PlayerLoginEvent;
+import org.eclipse.jdt.annotation.Nullable;
+import org.skriptlang.skript.lang.script.ScriptWarning;
+
 @Name("Cancel Event")
 @Description("Cancels the event (e.g. prevent blocks from being placed, or damage being taken).")
-@Examples({"on damage:",
-		"	victim is a player",
-		"	victim has the permission \"skript.god\"",
-		"	cancel the event"})
+@Examples({
+	"on damage:",
+		"\tvictim is a player",
+		"\tvictim has the permission \"skript.god\"",
+		"\tcancel the event"
+})
 @Since("1.0")
 public class EffCancelEvent extends Effect {
+
 	static {
 		Skript.registerEffect(EffCancelEvent.class, "cancel [the] event", "uncancel [the] event");
 	}
-	
+
 	private boolean cancel;
-	
-	@SuppressWarnings("null")
+
 	@Override
-	public boolean init(final Expression<?>[] vars, final int matchedPattern, final Kleenean isDelayed, final ParseResult parser) {
+	public boolean init(Expression<?>[] vars, int matchedPattern, Kleenean isDelayed, ParseResult parser) {
+		if (getParser().isCurrentEvent(PlayerLoginEvent.class))
+			Skript.error("A connect event cannot be cancelled, but the player may be kicked ('kick player by reason of \"...\"')");
 		if (isDelayed == Kleenean.TRUE) {
-			Skript.error("Can't cancel an event anymore after it has already passed", ErrorQuality.SEMANTIC_ERROR);
+			Skript.error("Can't cancel an event anymore after it has already passed");
 			return false;
 		}
 		cancel = matchedPattern == 0;
-		final Class<? extends Event>[] es = getParser().getCurrentEvents();
-		if (es == null)
+		Class<? extends Event>[] currentEvents = getParser().getCurrentEvents();
+		if (currentEvents == null)
 			return false;
-		for (final Class<? extends Event> e : es) {
-			if (Cancellable.class.isAssignableFrom(e) || BlockCanBuildEvent.class.isAssignableFrom(e))
-				return true; // TODO warning if some event(s) cannot be cancelled even though some can (needs a way to be suppressed)
+
+		int cancellable = 0;
+		for (Class<? extends Event> event : currentEvents) {
+			if (Cancellable.class.isAssignableFrom(event) || BlockCanBuildEvent.class.isAssignableFrom(event))
+				cancellable++;
 		}
-		if (getParser().isCurrentEvent(PlayerLoginEvent.class))
-			Skript.error("A connect event cannot be cancelled, but the player may be kicked ('kick player by reason of \"...\"')", ErrorQuality.SEMANTIC_ERROR);
-		else
-			Skript.error(Utils.A(getParser().getCurrentEventName()) + " event cannot be cancelled", ErrorQuality.SEMANTIC_ERROR);
+		// All events are cancellable.
+		if (cancellable == currentEvents.length)
+			return true;
+		// Some events are cancellable.
+		if (cancellable > 0) {
+			if (!getParser().getCurrentScript().suppressesWarning(ScriptWarning.VARIABLE_SAVE))
+				Skript.warning(Utils.A(getParser().getCurrentEventName()) + " can be called by multiple events, and some cannot be cancelled.");
```

Comment:
```suggestion
				Skript.warning(Utils.A(getParser().getCurrentEventName()) + " event can occur for multiple reasons, some of which cannot be cancelled.");
```
Something like this might sound better

Reply:
I like the ending, I don't know about the word `occur` over `called`

Changes require updates to the JUnit test string compare.

Reply:
Maybe the word trigger?

Reply:
ok